### PR TITLE
Wpi testing

### PR DESCRIPTION
**Pull request by BrownGenius**
These are the changes we made at WPI.  This code needs to be cleaned up prior to a push to main, but creating a PR to document the changes.

### CODE REVIEW

File: src/main/deploy/pathplanner/MobilityBlueHumanSideToDock.path [hunk 1]
```diff
@@ -70,15 +70,8 @@
...
-      "names": [
-        "event"
-      ]
-    }
-  ]
+  "maxVelocity": 2.0,
+  "maxAcceleration": 1.0,
+  "isReversed": null,
+  "markers": []
 }
```
Comment (BrownGenius) [inline src/main/deploy/pathplanner/MobilityBlueHumanSideToDock.path:77]: Updated autonomous routines to start from edge of community

File: src/main/deploy/robotconfig/00802F17DEE0.json [hunk 2]
```diff
@@ -24,31 +24,31 @@
...
+        "RIGHT_FEED_FORWARD_LINEAR_A": 0.67869,
 
-        "FEED_FORWARD_ANGULAR_S": 2,
-        "FEED_FORWARD_ANGULAR_V": 0.9143,
-        "FEED_FORWARD_ANGULAR_A": 4.5569
+        "FEED_FORWARD_ANGULAR_S": 1.1686,
+        "FEED_FORWARD_ANGULAR_V": 3.4539,
+        "FEED_FORWARD_ANGULAR_A": 1.6501
     }
 }
```
Comment (BrownGenius) [inline src/main/deploy/robotconfig/00802F17DEE0.json:54]: Updated based on sysId run @ WPI.  Not final.

File: src/main/java/bhs/devilbotz/Constants.java [hunk 3]
```diff
@@ -78,9 +78,9 @@ public static class DriveConstants {
...
   public static class GripperConstants {
     /** The gripper double solenoid forward channel */
-    public static final int GRIPPER_SOLENOID_FORWARD = 0;
+    public static final int GRIPPER_SOLENOID_FORWARD = 1;
     /** The gripper double solenoid reverse channel */
-    public static final int GRIPPER_SOLENOID_REVERSE = 1;
+    public static final int GRIPPER_SOLENOID_REVERSE = 2;
     /** The CAN ID for the gripper's pneumatic compressor */
     public static final int COMPRESSOR_CAN_ID = 6;
   }
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/Constants.java:86]: Gripper solenoids were reversed on the actual robot

File: src/main/java/bhs/devilbotz/RobotContainer.java [hunk 4]
```diff
@@ -67,6 +72,10 @@ public class RobotContainer {
 
   /** The container for the robot. Contains subsystems, OI devices, and commands. */
   public RobotContainer() {
+    if (Robot.isReal()) {
+      cameraManager = new CameraManager();
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/RobotContainer.java:76]: Only create cameraManager on real robot. _This should probably be handed inside of CameraManager_.

File: src/main/java/bhs/devilbotz/RobotContainer.java [hunk 5]
```diff
@@ -75,7 +84,6 @@ public RobotContainer() {
     SmartDashboard.putData(driveTrain);
 
     arm.setDefaultCommand(new ArmIdle(arm));
-    gripper.setDefaultCommand(new GripperIdle(gripper));
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/RobotContainer.java:78]: The default gripper command was always triggering, causing the gripper open/close command to fail.

File: src/main/java/bhs/devilbotz/RobotContainer.java [hunk 6]
```diff
@@ -93,11 +101,11 @@ public RobotContainer() {
...
 
     new JoystickButton(leftJoystick, 1)
-        .toggleOnTrue(new GripperClose(gripper))
+        .onTrue(new GripperClose(gripper))
         .onFalse(new GripperIdle(gripper));
 
     new JoystickButton(leftJoystick, 2)
-        .toggleOnTrue(new GripperOpen(gripper))
+        .onTrue(new GripperOpen(gripper))
         .onFalse(new GripperIdle(gripper));
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/RobotContainer.java:109]: @ParkerMeyers made this change. Not sure the difference between toggleOnTrue vs onTrue.

File: src/main/java/bhs/devilbotz/RobotContainer.java [hunk 7]
```diff
@@ -170,21 +179,27 @@ public Command getAutonomousCommand(AutonomousModes autoMode) {
...
+          if (Alliance.Blue == DriverStation.getAlliance()) {
+            path = PathPlanner.loadPath("MobilityBlueWallSideToDock", new PathConstraints(2.5, 1));
+          } else {
+            path = PathPlanner.loadPath("MobilityRedWallSideToDock", new PathConstraints(2.5, 1));
+          }
           autonomousCommand =
               Commands.waitSeconds(ShuffleboardManager.autoDelay.getDouble(0))
                   .asProxy()
-                  .andThen(driveTrain.followTrajectoryCommand(WallZoneMobilityDockAndEngage, true))
+                  .andThen(driveTrain.followTrajectoryCommand(path, true))
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/RobotContainer.java:202]: Added automatic selection of red vs blue paths.

File: src/main/java/bhs/devilbotz/commands/arm/ArmDown.java [hunk 8]
```diff
@@ -45,6 +45,6 @@ public void end(boolean interrupted) {
   // Returns true when the command should end.
   @Override
   public boolean isFinished() {
-    return arm.isBottomLimit() || arm.atBottom();
+    return arm.isBottomLimit();
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/commands/arm/ArmDown.java:48]: the `atBottom()` condition was causing the gripper arm to behave erratically when the bottom limit switch was broken/loose.

File: src/main/java/bhs/devilbotz/commands/gripper/GripperIdle.java [hunk 9]
```diff
@@ -43,6 +43,6 @@ public void end(boolean interrupted) {}
   // Returns true when the command should end.
   @Override
   public boolean isFinished() {
-    return false;
+    return true;
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/commands/gripper/GripperIdle.java:46]: @ParkerMeyers made this change.  Not sure why.

File: src/main/java/bhs/devilbotz/subsystems/Arm.java [hunk 10]
```diff
@@ -84,10 +84,22 @@ public class Arm extends SubsystemBase {
...
+   * low cone: 258
+   * middle cone: 468
+   * high cone: 558
+   * 
+   * cube
+   * low: 180
+   * middle: 354
+   * high: 440
+   * 
+   */
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/subsystems/Arm.java:102]: Empirical arm encoder values for cone vs cube scoring.

File: src/main/java/bhs/devilbotz/subsystems/DriveTrain.java [hunk 11]
```diff
@@ -577,7 +578,7 @@ private void updateOdometry() {
    * @since 1/30/2023
    */
   public double getRoll() {
-    return navx.getRoll();
+    return -navx.getRoll();
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/subsystems/DriveTrain.java:581]: navx on competition bot is rotated 180 compared to the practice bot, so the roll value needs to be negated. _This needs to be moved to the robotconfig json files as a flag.....e.g. rollInversion=true/false_

File: src/main/java/bhs/devilbotz/subsystems/DriveTrain.java [hunk 12]
```diff
@@ -600,7 +601,7 @@ public double getYaw() {
    *     Example</a>
    */
   public Command followTrajectoryCommand(PathPlannerTrajectory traj, boolean isFirstPath) {
-    field.getObject("path").setTrajectory(traj);
+    // field.getObject("path").setTrajectory(traj);
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/subsystems/DriveTrain.java:604]: In simulation, this code hides the actual robot location.  _should disable only when in simulation.....or make it transparent_

File: src/main/java/bhs/devilbotz/subsystems/Gripper.java [hunk 13]
```diff
@@ -32,16 +32,19 @@ public Gripper() {
...
   /** This method closes the gripper. */
   public void close() {
     doubleSolenoid.set(Value.kReverse);
+    System.out.println("Closing");
   }
 
   /** This method sets the grippers speed to 0. */
   public void stop() {
     doubleSolenoid.set(Value.kOff);
+    System.out.println("Stopping");
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/subsystems/Gripper.java:47]: Debug code for gripper

File: src/main/java/bhs/devilbotz/utils/ShuffleboardManager.java [hunk 14]
```diff
@@ -63,24 +63,24 @@ public ShuffleboardManager() {
...
-    autoModeChooser.addOption("Test", AutonomousModes.TEST);
+//    autoModeChooser.addOption(
+//        "Routine 7: Score, Dock, and Engage", AutonomousModes.SCORE_DOCK_AND_ENGAGE);
+//    autoModeChooser.addOption(
+//       "Routine 8: Score, Mobility, Dock, and Engage", AutonomousModes.SCORE_MOBILITY_DOCK_ENGAGE);
+//    autoModeChooser.addOption(
+//        "Routine 9: Score, Mobility, Pick, Dock, and Engage",
+//        AutonomousModes.SCORE_MOBILITY_PICK_DOCK_ENGAGE);
+//    autoModeChooser.addOption("Test", AutonomousModes.TEST);
     autoMode.add("Routine", autoModeChooser).withWidget(BuiltInWidgets.kComboBoxChooser);
```
Comment (BrownGenius) [inline src/main/java/bhs/devilbotz/utils/ShuffleboardManager.java:84]: Removed unimplemented routines.  Shortened naming.

File: src/main/java/bhs/devilbotz/subsystems/Gripper.java (hunk 13 above)
Comment (loafdog) [reply src/main/java/bhs/devilbotz/subsystems/Gripper.java:47]: I don't understand what `kOff` does for a solenoid.  We should figure that out and see if the stop method is needed.

File: src/main/java/bhs/devilbotz/RobotContainer.java (hunk 5 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/RobotContainer.java:78]: We can experiment with cleaning this up Monday when we have the actual robot.  The code as it is seems to do the correct thing.

File: src/main/java/bhs/devilbotz/RobotContainer.java (hunk 6 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/RobotContainer.java:102]: Yes, the gripper controls were not doing anything.  When the electrical team put a meter on the solenoid leads, they saw a short blip when the button was pressed, so something was prematurely turning the gripper solenoids, off.

File: src/main/java/bhs/devilbotz/commands/arm/ArmToTop.java [hunk 15]
```diff
@@ -43,6 +43,6 @@ public void end(boolean interrupted) {
   // Returns true when the command should end.
   @Override
   public boolean isFinished() {
-    return (arm.isTopLimit() || arm.atTop());
+    return (arm.isTopLimit());
```
[the remark this answers is not in the document]
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/commands/arm/ArmToTop.java:46]: Monday, we should pre-program/test the automatic arm height buttons.

File: src/main/java/bhs/devilbotz/commands/gripper/GripperIdle.java (hunk 9 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/commands/gripper/GripperIdle.java:46]: We can test other options on Monday.

File: src/main/java/bhs/devilbotz/subsystems/Gripper.java (hunk 13 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/subsystems/Gripper.java:47]: I believe stop turns off both solenoids, so keeps the gripper wherever it is currently. Neither the forward or reverse is actuated.

Comment (loafdog) [reply src/main/java/bhs/devilbotz/subsystems/Gripper.java:47]: Got it.  I think these prints should be removed.  They may fill up console with lots of msgs, or get lost if other print statements fill up console.  We should figure out how to put equivalent status to Shuffleboard.

**Review by loafdog — COMMENTED** [review]
One change I'd like to see, rest lgtm.

File: src/main/java/bhs/devilbotz/subsystems/Gripper.java (hunk 13 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/subsystems/Gripper.java]: Done. I removed the prints.

Comment (ParkerMeyers) [reply src/main/java/bhs/devilbotz/subsystems/Gripper.java]: yep they were just for quick debugging

File: src/main/java/bhs/devilbotz/Robot.java [hunk 16]
```diff
@@ -230,4 +233,12 @@ public static JsonNode getSysIdConstant(String name) {
   public static JsonNode getDriveTrainConstant(String name) {
     return robotConfig.get("drivetrain").get(name);
   }
+
+  public static boolean checkCapability(String name) {
```
Comment (ParkerMeyers) [inline src/main/java/bhs/devilbotz/Robot.java:237]: Javadoc please

File: src/main/java/bhs/devilbotz/Robot.java [hunk 17]
```diff
@@ -230,4 +233,12 @@ public static JsonNode getSysIdConstant(String name) {
   public static JsonNode getDriveTrainConstant(String name) {
```
Comment (ParkerMeyers) [inline src/main/java/bhs/devilbotz/Robot.java:233]: Javadocs, same with all other methods in Robot.json

File: src/main/java/bhs/devilbotz/RobotContainer.java [hunk 18]
```diff
@@ -93,11 +94,11 @@ public RobotContainer() {
   private void configureBindings() {
 
     new JoystickButton(leftJoystick, 1)
```
Comment (ParkerMeyers) [inline src/main/java/bhs/devilbotz/RobotContainer.java:96]: I know other bindings dont have it, but we should add comments to each binding in order to make it easier to understand the code

File: src/main/java/bhs/devilbotz/subsystems/DriveTrain.java [hunk 19]
```diff
@@ -289,6 +288,7 @@ public void resetRobotPosition() {
   public void periodic() {
     // Updates the odometry of the drive train.
     updateOdometry();
+    SmartDashboard.putNumber("Roll", navx.getRoll());
```
Comment (ParkerMeyers) [inline src/main/java/bhs/devilbotz/subsystems/DriveTrain.java:291]: Needed?

File: src/main/java/bhs/devilbotz/subsystems/DriveTrain.java [hunk 20]
```diff
@@ -460,14 +460,12 @@ private void setupTalons() {
...
-      // TODO: Understand why the sensor phase needs to be swapped for simulation
-      rightMaster.setSensorPhase(false);
-      leftMaster.setSensorPhase(false);
-    } else {
+    if (Robot.checkCapability("hasInvertedDriveEncoderSensorPhase")) {
       rightMaster.setSensorPhase(true);
       leftMaster.setSensorPhase(true);
+    } else {
+      rightMaster.setSensorPhase(false);
+      leftMaster.setSensorPhase(false);
```
Comment (ParkerMeyers) [inline src/main/java/bhs/devilbotz/subsystems/DriveTrain.java:468]: Do we know why the sensor phase needs to be changed?

**Review by ParkerMeyers — COMMENTED** [review]
Minor changes/Documentation.

Make sure everything has javadocs

**Comment by loafdog** [comment]
Suggested changes to add gripper to shuffleboard here: https://github.com/BHSRobotix/ChargedUp2023/tree/gripper_shuffleboard
I created the gripper_shuffleboard branch off of this one. I did not create a PR because not sure how/when to review it.  We can:
- create PR against this branch, review, and end up merging to this branch, and then merge this branch to main
- wait on creating PR/review until after this branch is merged to main
- ignore the branch if we do not want the changes to shuffleboard at all.

**Review by loafdog — APPROVED** [review]

**Comment by BrownGenius** [comment]
@loafdog I'm going to add the javadocs that Parker requested. I can pull in your branch to this and review it all together.

File: src/main/java/bhs/devilbotz/Robot.java (hunk 16 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/Robot.java:243]: done

File: src/main/java/bhs/devilbotz/Robot.java (hunk 17 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/Robot.java:233]: already had javadocs

File: src/main/java/bhs/devilbotz/RobotContainer.java (hunk 18 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/RobotContainer.java:96]: I agree, we should add comments as part of a separate PR.

File: src/main/java/bhs/devilbotz/subsystems/DriveTrain.java (hunk 20 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/subsystems/DriveTrain.java:467]: My understanding is the kitbot drivetrain doesn't have a fixed "front" vs "back", so the encoder phase depends on how they are mounted and what we decide "forward" on the robot is.  According to the docs, the phase needs to be determined empirically by driving the robot forward and set the phase such that the both left and right encoder values are positive.

**Review by loafdog — APPROVED** [review]
Thanks for doing all this work,  and for merging in shuffleboard changes.

**Pull request merged**